source/globalCommands.py

```diff
@@ -2023,29 +2023,30 @@ def script_review_currentSymbol(self,gesture):
 	@script(
 		description=_(
 			# Translators: Input help mode message for cycle speech mode command.
-			"Cycles between the speech modes of off, beep, talk and on-demand."
+			"Cycles between speech modes."
 		),
 		category=SCRCAT_SPEECH,
 		gesture="kb:NVDA+s"
 	)
 	def script_speechMode(self, gesture: inputCore.InputGesture) -> None:
 		curMode = speech.getState().speechMode
 		speech.setSpeechMode(speech.SpeechMode.talk)
-		newMode = (curMode + 1) % len(speech.SpeechMode)
-		if newMode == speech.SpeechMode.off:
-			# Translators: A speech mode which disables speech output.
-			name=_("Speech mode off")
-		elif newMode == speech.SpeechMode.beeps:
-			# Translators: A speech mode which will cause NVDA to beep instead of speaking.
-			name=_("Speech mode beeps")
-		elif newMode == speech.SpeechMode.talk:
-			# Translators: The normal speech mode; i.e. NVDA will talk as normal.
-			name=_("Speech mode talk")
-		elif newMode == speech.SpeechMode.onDemand:
-			# Translators: The on-demand speech mode; i.e. NVDA will talk only on commands asking to report something.
-			name = _("Speech mode on-demand")
+		modesList = list(speech.SpeechMode)
+		currModeIndex = modesList.index(curMode)
+		excludedModesIndexes = config.conf["speech"]["excludedSpeechModes"]
+		possibleIndexes = [i for i in range(len(modesList)) if i not in excludedModesIndexes]
+		# Sort speech modes to present next modes in the list before the ones at the beginning of the list.

```

Comment:
Thanks for the clarification in the detailed explanation. This makes things much more understandable IMO. Good job!